cmd/testscript/main.go

```diff
@@ -0,0 +1,249 @@
+		if err := os.Mkdir(rd, 0777); err != nil {
+			return fmt.Errorf("failed to create a run directory within %v for %v: %v", td, fn, err)
+		}
+		if err := run(rd, fn); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+var failedRun = errors.New("failed run")
+
+type runner struct {
+	verbose bool
+}
+
+func (r runner) Skip(is ...interface{}) {
+	panic("Skip is not supported")
+}
+
+func (r runner) Fatal(is ...interface{}) {
+	r.Log(is...)
+	r.FailNow()
+}
+
+func (r runner) Parallel() {
+	// No-op for now; we are currently only running a single script in a
+	// testscript instance.
+}
+
+func (r runner) Log(is ...interface{}) {
+	fmt.Print(is...)
+}
+
+func (r runner) FailNow() {
+	panic(failedRun)
+}
+
+func (r runner) Run(n string, f func(t testscript.T)) {
+	// For now we we don't top/tail the run of a subtest. We are currently only
+	// running a single script in a testscript instance, which means that we
+	// will only have a single subtest.
+	f(r)
+}
+
+func (r runner) Verbose() bool {
+	return r.verbose
+}
+
+func run(rd, fn string) error {
+	var ar *txtar.Archive
+	var err error
+
+	testdata := filepath.Join(rd, "testdata")
+	mods := filepath.Join(testdata, "goproxytest")
+
+	if err := os.MkdirAll(mods, 0777); err != nil {
+		return fmt.Errorf("failed to create testdata dir: %v", err)
+	}
+
+	if fn == os.Stdin.Name() {
+		byts, err := ioutil.ReadAll(os.Stdin)
+		if err != nil {
+			return fmt.Errorf("failed to read from stdin: %v", err)
+		}
+		ar = txtar.Parse(byts)
+	} else {
+		ar, err = txtar.ParseFile(fn)
+	}
+
+	if err != nil {
+		return fmt.Errorf("failed to txtar parse %v: %v", fn, err)
+	}
+
+	var toWrite []txtar.File
+
+	for _, f := range ar.Files {
+		fp := filepath.Clean(filepath.FromSlash(f.Name))
+		if isAbs(fp) || strings.HasPrefix(fp, ".."+string(filepath.Separator)) {
+			return fmt.Errorf("file %v would result in file outside parent directory", f.Name)
+		}
+
+		// only write out goproxytest rooted files
+		if strings.Split(fp, string(os.PathSeparator))[0] != "goproxytest" {
+			toWrite = append(toWrite, f)
+			continue
+		}
+
+		fp = filepath.Join(testdata, fp)
+
+		if err := os.MkdirAll(filepath.Dir(fp), 0777); err != nil {
+			return err
+		}
+		// Avoid overwriting existing files by using O_EXCL.
+		out, err := os.OpenFile(fp, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0666)
+		if err != nil {
+			return err
+		}
+
+		// unquote, because this will be a quoted goproxytest file
+		data := f.Data
+		data = bytes.Replace(data, []byte("\n>"), []byte("\n"), -1)
+		data = bytes.TrimPrefix(data, []byte(">"))
+
+		_, err = out.Write(data)
+		cerr := out.Close()
+		if err != nil {
+			return err
+		}
+		if cerr != nil {
+			return cerr
+		}
+	}
+
+	ar.Files = toWrite
+
+	if err := ioutil.WriteFile(filepath.Join(testdata, "script.txt"), txtar.Format(ar), 0666); err != nil {
+		return fmt.Errorf("failed to write script for %v: %v", fn, err)
+	}
+
+	srv, err := goproxytest.NewServer(mods, "")
```

Comment:
This should be shut down after the test has run, I think.

Also, perhaps don't start the proxy when there are no goproxytest files?

Reply:
> This should be shut down after the test has run, I think.

Good point. It also caught the fact that `Close()` is buggy 😄 

> Also, perhaps don't start the proxy when there are no goproxytest files?

Done.